kernel/src/hil/rng.rs

```diff
@@ -66,16 +67,32 @@ pub enum Continue {
     Done,
 }
 
-/// Generic interface for a random number generator
+/// Generic interface for random number generation.
 ///
 /// Implementors should assume the client implements the
-/// [Client](trait.Client.html) trait.
-pub trait RNG {
-    /// Initiate the aquisition of new random number generation.
+/// [Client](trait.Client.html) trait. Note that the random
+/// bits generated do not have entropy guarantees. Therefore,
+/// a caller who needs strong entropy (e.g., for cryptography)
+/// should request many more bits than needed and compress their
+/// entropy with a cryptographic hash and should do so with a
+/// good understanding of the true underlying entropy provided.
+///
+/// Implementers of this
+/// trait should try to ensure that there is at least 0.25 bits of
+/// entropy per generated bit, such that sampling 4x the bits needed
+/// can generate sufficient entropy, but cryptosystems should verify
+/// this is true for each implementation they rely on.
+
+pub trait RNG<'a> {
+    /// Request new random bits.
     ///
-    /// The implementor may ignore this command if the generation proccess is
-    /// already in progress.
+    /// This method will cause `randomness_available` being called on the Client
+    /// at some point soon in the future. Numerious calls to this method before
+    /// `randomness_available` is called will only result in one callback.
```

Comment:
Should a repeat call perhaps instead return `EALREADY`? Keeps a one call -> one callback pattern?

Reply:
Yes -- a return code would be great.

Reply:
Actually, wait. Why should there be a one call -> one callback pattern? If the semantics are "SUCCESS means you will get a callback" then EALREADY implies you will not get a callback. EALREADY is for cases of actuation (you are changing something) rather than acquisition. It's a very useful idiom to start something if it isn't started and if it is started just keep on going. E.g., you want to turn on a peripheral: trying to turn it on several times is OK.

Reply:
Should it perhaps then be an error to change the client if there is an outstanding callback pending?

I agree that it's reasonable for a continuation interface to allow repeated start calls. The motivation for the one -> one pattern in my mind is trying to ensure the contract of "if you get a `SUCCESS` then your callback *will* be called"

Reply:
NB: I've moved a larger discussion to #1069. I'll close this PR.

I agree -- if you get a SUCCESS then your callback will be called. However, this is a many -> one pattern, because there's only a single callback. Otherwise there's implicit queueing.

An error code for a client change is an interesting thought. I'm not sure this is always an error, though. For example, what if you have a virtualizer that's trying to feed randomness to multiple clients. It might, in the randomness_available callback, change the client before returning Continue::More. Or, if a higher priority client request comes in, and it uses a helper structure to handle the callbacks, it might want to change the callback for an outstanding acquisition. So while I think that there are cases when changing the client mid-acquisition would be a bug, it could also be the correct behavior; we probably don't want the implementation to make assumptions about what its caller does.